### ISSUE TITLE
can't run

### ISSUE THREAD

**Issue by mingxin-yang** [pinterest/api-quickstart#33]

**Comment by davidchaiken** [comment]
Hi, @mingxin-yang. We're happy to help with issues that you're having with this quickstart, but obviously need more information to be able to provide assistance.

**Comment by mingxin-yang** [comment]
I use node to run. this is the error:
```bash
localhost:nodejs fourmol$ ./scripts/get_access_token.js 
Using application ID and secret from PINTEREST_APP_ID and PINTEREST_APP_SECRET.
reading access_token_v5 from environment failed, trying read
reading access_token_v5 failed, trying oauth
v5 requires scopes for OAuth. setting to default: READ_USERS,READ_PINS,READ_BOARDS
getting auth_code...
exchanging auth_code for access_token...
POST https://api.pinterest.com/v5/oauth/token
TypeError: Cannot read property 'body' of undefined
    at AccessToken.print_and_throw_error (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/src/api_common.js:40:73)
    at AccessToken.oauth (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/src/v5/access_token.js:67:12)
    at processTicksAndRejections (node:internal/process/task_queues:94:5)
    at async AccessToken.fetch (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/src/access_token_common.js:54:5)
    at async main (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/scripts/get_access_token.js:71:7)
usage: get_access_token.js [-h] [-w] [-ct] [-s SCOPES] [-a ACCESS_TOKEN]
                           [-l LOG_LEVEL] [-v API_VERSION]
```

**Comment by mingxin-yang** [comment]
@davidchaiken

**Comment by davidchaiken** [comment]
Thanks for this report! I’ll try to reproduce the issue.

**Comment by mingxin-yang** [comment]
Thank you for your reply.
1. below is my environment
> macOS 10.13.6
> node v15.6.0
2. Also reported an error when getting `access_token`
3. this is working

After this, I think it may be a problem with the node version, I will try something else

**Comment by davidchaiken** [comment]
i could not reproduce the problem using node version 15.6.0. what shell are you using? (what is that output of `echo $0` at the shell prompt?)

thanks for trying the tutorial. that verifies that your application id and secret are correct.

i would appreciate it if you would post the output of this command, which should remove all of the confidential material from the log: `./scripts/get_access_token.js --log-level 3 | grep -E -v '_token:|code:|code='`

**Comment by mingxin-yang** [comment]
> $ echo $0
-bash   

> $ ./scripts/get_access_token.js --log-level 3 | grep -E -v '_token:|code:|code='
Using application ID and secret from PINTEREST_APP_ID and PINTEREST_APP_SECRET.
reading access_token_v5 from environment failed, trying read
  because... Error: No access token in the environment
    at AccessToken.from_environment (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/src/access_token_common.js:73:13)
    at AccessToken.fetch (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/src/access_token_common.js:35:12)
    at main (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/scripts/get_access_token.js:71:26)
reading access_token_v5 failed, trying oauth
  because... Error: ENOENT: no such file or directory, open 'access_token_v5.json'
    at Object.openSync (node:fs:585:3)
    at Object.readFileSync (node:fs:453:35)
    at AccessToken.read (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/src/access_token_common.js:79:32)
    at AccessToken.fetch (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/src/access_token_common.js:45:12)
    at main (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/scripts/get_access_token.js:71:26) {
  errno: -2,
  syscall: 'open',
  path: 'access_token_v5.json'
}
v5 requires scopes for OAuth. setting to default: READ_USERS,READ_PINS,READ_BOARDS
getting auth_code...
OAuth URI: https://www.pinterest.com/oauth/?consumer_id=1474276&redirect_uri=http://localhost:8085/&response_type=code&refreshable=true&scope=user_accounts:read,pins:read,boards:read&state=75135e37-d17d-4917-9f62-170bfb15ff0e
WARNING: This log has clear text credentials that need to be protected.
exchanging auth_code for access_token...
POST https://api.pinterest.com/v5/oauth/token
WARNING: This log has clear text credentials that need to be protected.
{
  redirect_uri: 'http://localhost:8085/',
  grant_type: 'authorization_code'
}
TypeError: Cannot read properties of undefined (reading 'body')
    at AccessToken.print_and_throw_error (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/src/api_common.js:40:73)
    at AccessToken.oauth (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/src/v5/access_token.js:66:12)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async AccessToken.fetch (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/src/access_token_common.js:54:5)
    at async main (file:///Users/fourmol/AndroidDev/files/api-quickstart-main/nodejs/scripts/get_access_token.js:71:7)
usage: get_access_token.js [-h] [-w] [-ct] [-s SCOPES] [-a ACCESS_TOKEN]
                           [-l LOG_LEVEL] [-v API_VERSION]

**Comment by davidchaiken** [comment]
Thank you again for this problem report, @mingxin-yang. It appears that the error handling code in the quickstart is not written robustly enough to print out a meaningful error to diagnose the problem that you are seeing.

I have pushed a version of the code with better error handling to a branch on this repo. I would appreciate it if you would check out this branch with `git checkout origin/dc_issue_33` and then try to run the command. You should receive a more understandable error message.

It's getting late in the US, so I won't respond again before tomorrow, but will check for your response in the morning. Thank you very much for your patience!

**Comment by mingxin-yang** [comment]
It's up and running, because Pinterest is blocked by China, I can't request to use `got`, after add agent to `got`, It work.
```
import HttpsProxyAgent from 'https-proxy-agent';
var agent = new HttpsProxyAgent('http://127.0.0.1:8001');

const response = await got.post(${this.api_uri}/v5/oauth/token, {

  agent: {

    https: agent,

    http: agent

  },

  headers: this.auth_headers, // use the recommended authorization approach

  form: post_data, // send body as x-www-form-urlencoded

  responseType: 'json'

});
```

Thank you for your patient reply, I'm really sorry that it took you a long time.

**Comment by davidchaiken** [comment]
Thanks for the update @mingxin-yang. Every issue report is a learning opportunity, so I appreciate getting the chance to work with you. Here are a couple of questions:

1. Did the error logging that I added to the quickstart help you understand the issue?
2. If we added support for the HttpsProxyAgent to the quickstart, would you be willing to test that it works for you?

**Comment by mingxin-yang** [comment]
1. Yes, the error logging help me. Printed out the `connect etimedout` problem.
2. of course, that would be better.

I have another suggestion.  How to request data in the v5 document, for example, `list boards`. It is not clearly written about the substitution of `access_token`, but it is written in the v3 document. I think that is better.
https://developers.pinterest.com/docs/api/v5/#tag/boards
https://developers.pinterest.com/docs/redoc/martech/#operation/v3_user_profile_boards_feed_GET

**Comment by davidchaiken** [comment]
Thank you very much for the feedback. I will release the error logging soon, and maybe implement the httpproxyagent in the next sprint. It would be helpful to know which proxy agent you are using so that I can test with that proxy.

Thanks also for the feedback on documentation. Are you referring to this part of the v3 documentation? `-H 'Authorization: Bearer <Your API token>'`

**Comment by mingxin-yang** [comment]
> It would be helpful to know which proxy agent you are using so that I can test with that proxy.

Does `proxy agent` mean software? I use V2RayX on my mac computer.

> Are you referring to this part of the v3 documentation? 

Yes, I don't see this in v5

**Comment by davidchaiken** [comment]
Thanks again for your feedback, @mingxin-yang. Based on your feedback:
* I merged improved error handling in this repo for both nodejs and python.
* I passed along your feedback on documentation to an documentation expert at Pinterest. The development team will add your request to document each endpoint that requires an access token to our product backlog.
* I have added your request to support HttpsProxyAgent to our product backlog.

Best wishes for success building an application on the Pinterest API!
